servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/SingleConcatWithPublisher.java

```diff
@@ -43,100 +47,144 @@ Executor executor() {
+            mayBeResult = CANCELLED;
+            super.cancel();
+        }
+
+        /**
+         * Helper method to invoke {@link DelayedCancellableThenSubscription#cancel()} from subclasses.
+         */
+        final void superCancel() {
+            super.cancel();
+        }
+
+        final boolean tryEmitSingleSuccessToTarget(@Nullable final T result) {
+            try {
+                target.onNext(result);
+                return true;
+            } catch (Throwable cause) {
+                mayBeResult = CANCELLED;
+                target.onError(cause);
+                return false;
+            }
+        }
+    }
+
+    private static final class ConcatSubscriber<T> extends AbstractConcatSubscriber<T> {
+        /**
+         * If {@link #request(long)} (with a valid n) invoked before {@link #onSuccess(Object)}.
+         */
+        private static final Object REQUESTED = new Object();
+
+        ConcatSubscriber(final Subscriber<? super T> target, final Publisher<? extends T> next) {
+            super(target, next);
+        }
+
+        @Override
+        public void onSuccess(@Nullable final T result) {
+            for (;;) {
+                final Object oldValue = mayBeResult;
+                if (oldValue == REQUESTED) {
+                    if (tryEmitSingleSuccessToTarget(result)) {
+                        next.subscribeInternal(this);
+                    }
+                    break;
+                } else if (oldValue == CANCELLED || mayBeResultUpdater.compareAndSet(this, INITIAL, result)) {
+                    break;
+                }
+            }
         }
 
         @Override
         public void request(long n) {
             for (;;) {
-                Object oldVal = mayBeResult;
-                if (oldVal == REQUESTED || oldVal == CANCELLED) {
+                final Object oldVal = mayBeResult;
+                if (oldVal == CANCELLED) {
+                    break;
+                } else if (oldVal == REQUESTED) {
                     super.request(n);
                     break;
                 } else if (!isRequestNValid(n)) {
                     mayBeResult = CANCELLED;
                     try {
                         target.onError(newExceptionForInvalidRequestN(n));
                     } finally {
-                        super.cancel();
+                        superCancel();
                     }
                     break;
                 } else if (mayBeResultUpdater.compareAndSet(this, oldVal, REQUESTED)) {
                     if (oldVal != INITIAL) {
                         @SuppressWarnings("unchecked")
                         final T tVal = (T) oldVal;
-                        emitSingleSuccessToTarget(tVal);
+                        if (tryEmitSingleSuccessToTarget(tVal)) {
+                            next.subscribeInternal(this);
+                        }
                     }
                     if (n != 1) {
                         super.request(n - 1);
```

Comment:
FYI: I observed this issue with pre-existing behavior of ``: https://github.com/apple/servicetalk/runs/2915774996
It does not happen if I move this `super.request(n - 1)` before `if (oldVal != INITIAL)`.

Reply:
Changed order in https://github.com/apple/servicetalk/pull/1643/commits/dc592f2d5d065d0945e8a085bbad247d1c9edee2

Reply:
Issue was fixed in #1646